activerecord/lib/active_record/asynchronous_queries_tracker.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+require "weakref"
+
+module ActiveRecord
+  class AsynchronousQueriesTracker # :nodoc:
```

Comment:
Instead of holding a weakref to all the queued async queries, we could probably give there queries a reference to the tracker (maybe rename it?) and have an AtomicBoolean to represent whether the query is cancelled.

Reply:
I'm not sure I follow. You need to be able to iterate over all the queries a request (or job) generated to be able to cancel them if the request is interrupted.

How would you do this with an inverted reference?

Reply:
Cancelling is just an ivar (there's no way to actually cancel a query in-flight, and we probably wouldn't want to), so I imagined just checking whether our current "context" was cancelled when we hit `execute` (I don't believe there is harm in "cancelled" queries sitting in the queue until they noop).

Reply:
> I imagined just checking whether our current "context" was cancelled when we hit `execute`

Oh I see. That makes sense. 👍